ssl/quic/quic_record_rx.c

```diff
@@ -253,6 +271,209 @@ void ossl_qrx_inject_urxe(OSSL_QRX *qrx, QUIC_URXE *urxe)
+ * initial packet only. All other packets should be discarded. This allows
+ * port_default_packet_handler() to validate incoming packet. If packet
+ * is not valid, then port_default_packet_handler() must discard the
+ * packet instead of creating a new channel for it.
+ */
+static int qrx_validate_initial_pkt(OSSL_QRX *qrx, QUIC_URXE *urxe,
+                                    const QUIC_CONN_ID *first_dcid,
+                                    size_t datagram_len)
+{
+    PACKET pkt, orig_pkt;
+    RXE *rxe;
+    size_t i = 0, aad_len = 0, dec_len = 0;
+    const unsigned char *sop;
+    unsigned char *dst;
+    QUIC_PKT_HDR_PTRS ptrs;
+    uint32_t pn_space;
+    OSSL_QRL_ENC_LEVEL *el = NULL;
+    uint64_t rx_key_epoch = UINT64_MAX;
+
+    if (!PACKET_buf_init(&pkt, ossl_quic_urxe_data(urxe), urxe->data_len))
+        return 0;
+
+    orig_pkt = pkt;
+    sop = PACKET_data(&pkt);
+
+    /*
+     * Get a free RXE. If we need to allocate a new one, use the packet length
+     * as a good ballpark figure.
+     */
+    rxe = qrx_ensure_free_rxe(qrx, PACKET_remaining(&pkt));
+    if (rxe == NULL)
+        return 0;
+
+    /*
+     * we expect INITIAL packet only, therefore it is OK to pass
+     * short_conn_id_len as 0.
+     */
+    if (!ossl_quic_wire_decode_pkt_hdr(&pkt,
+                                       0, /* short_conn_id_len */
+                                       1, /* need second decode */
+                                       0, /* nodata -> want to read data */
+                                       &rxe->hdr, &ptrs,
+                                       NULL))
+        goto malformed;
+
+    if (rxe->hdr.type != QUIC_PKT_TYPE_INITIAL)
+        goto malformed;
+
+    if (!qrx_validate_hdr_early(qrx, rxe, NULL))
+        goto malformed;
+
+    if (ossl_qrl_enc_level_set_have_el(&qrx->el_set, QUIC_ENC_LEVEL_INITIAL) != 1)
+        goto malformed;
+
+    if (rxe->hdr.type == QUIC_PKT_TYPE_INITIAL) {
+        const unsigned char *token = rxe->hdr.token;
+
+        /*
+         * This may change the value of rxe and change the value of the token
+         * pointer as well. So we must make a temporary copy of the pointer to
+         * the token, and then copy it back into the new location of the rxe
+         */
+        if (!qrx_relocate_buffer(qrx, &rxe, &i, &token, rxe->hdr.token_len))
+            goto malformed;
+
+        rxe->hdr.token = token;
+    }
+
+    pkt = orig_pkt;
+
+    el = ossl_qrl_enc_level_set_get(&qrx->el_set, QUIC_ENC_LEVEL_INITIAL, 1);
+    assert(el != NULL); /* Already checked above */
+
+    if (!ossl_quic_hdr_protector_decrypt(&el->hpr, &ptrs))
+        goto malformed;
+
+    /*
+     * We have removed header protection, so don't attempt to do it again if
+     * the packet gets deferred and processed again.
+     */
+    pkt_mark(&urxe->hpr_removed, 0);
+
+    /* Decode the now unprotected header. */
+    if (ossl_quic_wire_decode_pkt_hdr(&pkt, 0,
```

Comment:
```suggestion
    if (!ossl_quic_wire_decode_pkt_hdr(&pkt, 0,
```
Also don't forget to indent line below